File: Foundation/orbsupport/java/src/cdmw/orbsupport/StrategyList.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.orbsupport;

import cdmw.common.BadParameterException;

/**
 * This class defines is used to set ORB dependent
 * capability in an ORB independent way.
 *
 * Features:
 * Thread safe, exception safe
 *
 * @see  ORBSupport
 *
 */
public class StrategyList {

    /**
     * If true, the user has asked for a multi-threaded ORB.
     */
    private boolean hasORBThreaded;

    /**
     * If true, the user has asked for a Threaded POA
     */
    private boolean hasPOAThreadPerRequest;

    /**
     * If true, the user has asked for a Threaded POA
     */
    private boolean hasPOAThreadPerConnection;

    /**
     * If true, the user has asked for a POA with Thread Pool
     */
    private boolean hasPOAThreadPool;

    /**
     * this variable contains the thread number to be used by the ORB.
     */
    private int nbPOAThreads;

    /**
     * If true, the user has asked to use a specific port
     * for the POA
     */
    private boolean hasPOAPortAssigned;


    /**
     * Value of the port to be used for the POA, if
     * hasPOAPortAssigned has been set.
     */
    private int poaPort;

    /**
     * Constructor
     *
     */
    public StrategyList() {

        hasORBThreaded = false;
        hasPOAThreadPerRequest = false;
        hasPOAThreadPerConnection = false;
        hasPOAThreadPool = false;
        nbPOAThreads = 0;
        hasPOAPortAssigned = false;
        poaPort = 0;

    }

    /**
     * Set the ORB shall use multi-threading.
     *
     */
    public void addORBThreaded() {

        this.hasORBThreaded = true;

    }

    /**
     * Set the POA shall use multi-threading. The threads
     *  will be created at the initialisation time. 'nbThreads'
     *  will be created.
     *
     *  This Strategy is incompatible with:
     *   - addPOAThreadPerRequest
     *   - addPOAThreadPerConnection
     *
     *  This Strategy requires:
     *   - ORBInit() call with addORBThreaded Strategy
     *
     * @param nbThreads the size of the thread pool
     *
     * @exception BadParameterException if the nbThreads is equal to zero
     * @exception BadParameterException if the Strategy is incompatible
     *           with a previous one
     */
    public void addPOAThreadPool(int nbThreads)
        throws BadParameterException {

        // Check incompatible Strategy
        if ( hasPOAThreadPerRequest ) {
            throw new BadParameterException("POAThreaded already set");
        }

        if ( hasPOAThreadPerConnection ) {
            throw new BadParameterException("POAThreaded already set");
        }

        // Check already set Strategy
        if ( hasPOAThreadPool ) {
            throw new BadParameterException("POAThreadPool already set");
        }

        // Check Parameters
        if ( nbThreads > 0 ) {
            hasPOAThreadPool = true;
            nbPOAThreads = nbThreads;
        } else {
            throw new BadParameterException("minimum value of nbThreads is 1");
        }

    }

    /**
     * Set the POA shall use multi-threading: one thread per request
     *
     *  This Strategy is incompatible with:
     *   - addPOAThreadPerConnection
     *   - addPOAThreadPool
     *
     *  This Strategy requires:
     *   - ORBInit() call with addORBThreaded Strategy
     *
     * @exception BadParameterException if the Strategy is incompatible
     *           with a previous one
     */
    public void addPOAThreadPerRequest() throws BadParameterException {

        // Check incompatible Strategy
        if ( hasPOAThreadPool ) {
            throw new BadParameterException("POAThreadPool already set");
        }

        if ( hasPOAThreadPerConnection ) {
            throw new BadParameterException("POAThreaded already set");
        }

        // Check already set Strategy
        if ( hasPOAThreadPerRequest ) {
            throw new BadParameterException("POAThreaded already set");
        }

        // Check Parameters
        hasPOAThreadPerRequest = true;

    }

    /**
     * Set the POA shall use multi-threading: one thread per request
     *
     *  This Strategy is incompatible with:
     *   - addPOAThreadPool
     *   - addPOAThreadPerRequest
     *
     *  This Strategy requires:
     *   - ORBInit() call with addORBThreaded Strategy
     *
     * @exception BadParameterException if the Strategy is incompatible
     *           with a previous one
     */
    public void addPOAThreadPerConnection() throws BadParameterException {

        // Check incompatible Strategy
        if ( hasPOAThreadPool ) {
            throw new BadParameterException("POAThreadPool already set");
        }

        if ( hasPOAThreadPerRequest ) {
            throw new BadParameterException("POAThreaded already set");
        }

        // Check already set Strategy
        if ( hasPOAThreadPerConnection ) {
            throw new BadParameterException("POAThreaded already set");
        }

        // Check Parameters
        hasPOAThreadPerConnection = true;

    }

    /**
     * Returns if the Multi-threaded strategy has been
     * choosen for the ORB
     *
     * @return true if the strategy has been set, false elsewhere
     */
     public boolean isORBThreaded() {
         return hasORBThreaded;
     }

    /**
     * Returns true if the Thread Pool strategy has been
     * choosen for the POA
     *
     * @return true if the strategy has been set, false elsewhere
     */
    public boolean isPOAThreadPool() {
        return hasPOAThreadPool;
    }

    /**
     * Returns true if the ThreadPerConnection strategy has been
     * choosen for the POA
     *
     * @return true if the strategy has been set, false elsewhere
     */
    public boolean isPOAThreadPerConnection() {
        return hasPOAThreadPerConnection;
    }

    /**
     * Returns the number of threads of the POA
     *
     * @return threads number of the POA
     */
    public int getNbPOAThreads() {
        return this.nbPOAThreads;
    }

    /**
     * Returns true if the ThreadPerRequest strategy has been
     * choosen for the POA.
     *
     * @return true if the strategy has been set, false elsewhere
     */
    public boolean isPOAThreadPerRequest() {
        return hasPOAThreadPerRequest;
    }


   /**
    * Returns true if the user has asked for a specific
    * port for one POA
    *
    * @deprecated This should not be used by application
    * developpers as it will be removed in the next release.
    *
    * @return true if the strategy has been set, false elsewhere
    */
    protected boolean isPOAPortAssigned() {
        return hasPOAPortAssigned;
    }

    /**
     * Returns the port of the POA
     *
     * @return port of the POA
     */
    protected int getPOAPort() {
        return this.poaPort;
    }

    /**
     * Define the POA must use the specified port for
     * managing incoming CORBA request.
     *
     * @deprecated This should not be used by application
     * developpers as it will be removed in the next release.
     *
     * @param port the port to be used ( 0 < port < 65535 )
     *
     * @exception BadParameterException if the Strategy is incompatible
     *          with a previous one
     * @exception BadParameterException if the port number
     *            is not in the allowed range.
     */
    protected void addPOAPortAssigned(int poaPort)
        throws BadParameterException {


        // Check incompatible Strategy
        // None

        // Check already set Strategy
        if ( hasPOAPortAssigned ) {
            throw new BadParameterException("POAPortAssigned already set");
        }

        // Check Parameters
        if ( ! (poaPort > 0 && poaPort < 65535)  ) {
            throw new
                BadParameterException("The Port is not in the allowed range");
        }

        this.hasPOAPortAssigned = true;
        this.poaPort = poaPort;

    }

    /**
     * Copy the current StrategyList object
     *
     * @return copy of the current object
     */
    public StrategyList copy() {

        StrategyList copy = new StrategyList();

        if ( this.isORBThreaded() ) {
            copy.addORBThreaded();
        }
        if ( this.isPOAThreadPerRequest() ) {
            try {
                copy.addPOAThreadPerRequest();
            } catch(BadParameterException bpe) {}
        }
        if ( this.isPOAThreadPool() ) {
            try {
                copy.addPOAThreadPool( this.getNbPOAThreads() );
            } catch(BadParameterException bpe) {}
        }
        if ( this.isPOAPortAssigned() ) {
            try {
                copy.addPOAPortAssigned( this.getPOAPort() );
            } catch(BadParameterException bpe) {}
        }

        return copy;
    }

}

File: trunk/Foundation/orbsupport/c++/src/StrategyList.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include "Foundation/orbsupport/StrategyList.hpp"

#include <algorithm>

namespace Cdmw
{

    namespace OrbSupport
    {

        /**
        *Implementation notes:  none
        *<p>
        *Portability issues: none 
        *<p>
        */


        StrategyList::StrategyList()
        throw()
                : m_has_orbThreaded(false),

                m_has_poaThreadPerRequest(false),
                m_has_poaThreadPerConnection(false),
                m_has_poaThreadPool(false),
                m_nbPoaThreads(0),
                m_has_multicast(false),
                m_has_poaPortAssigned(false),
                m_poaPort(0)
        {
            // Nothing to do
        }


        StrategyList::StrategyList( const StrategyList& rhs )
        throw()
                :
                // Orb thread pool
                m_has_orbThreaded( rhs.m_has_orbThreaded ),

                //  Poa threaded
                m_has_poaThreadPerRequest( rhs.m_has_poaThreadPerRequest ),
                m_has_poaThreadPerConnection( rhs.m_has_poaThreadPerConnection ),

                //  Poa thread pool
                m_has_poaThreadPool( rhs.m_has_poaThreadPool ),
                m_nbPoaThreads( rhs.m_nbPoaThreads ),

                // Multicast capability
                m_has_multicast( rhs.m_has_multicast ),

                // Poa port
                m_has_poaPortAssigned( rhs.m_has_poaPortAssigned ),
                m_poaPort( rhs.m_poaPort )

        {

            // Nothing to do

        }


        //
        // Assignement operator
        //
        StrategyList&
        StrategyList::operator=( const StrategyList& rhs )
        throw()
        {

            StrategyList temp( rhs );
            swap ( temp );

            return *this;
        }



        //
        // Dtor
        //
        StrategyList::~StrategyList()
        throw()
        {
            // Nothing to do
        }



        //
        // Swap both objects
        //
        void
        StrategyList::swap( StrategyList& rhs )
        throw()
        {
            // Orb threaded
            std::swap( m_has_orbThreaded, rhs.m_has_orbThreaded );

            // Poa threaded
            std::swap( m_has_poaThreadPerRequest, rhs.m_has_poaThreadPerRequest );
            std::swap( m_has_poaThreadPerConnection, rhs.m_has_poaThreadPerConnection );

            // Poa thread pool
            std::swap( m_has_poaThreadPool, rhs.m_has_poaThreadPool );
            std::swap( m_nbPoaThreads, rhs.m_nbPoaThreads );

            // Multicast capability
            std::swap( m_has_multicast, rhs.m_has_multicast );

            // Poa port
            std::swap( m_has_poaPortAssigned, rhs.m_has_poaPortAssigned );
            std::swap( m_poaPort, rhs.m_poaPort );

        }



        //
        // Defines the threaded startegy for the ORB
        //
        void
        StrategyList::add_OrbThreaded()
        throw( BadParameterException )
        {
            m_has_orbThreaded = true;
        }


        //
        // Defines the Thread Pool strategy for the POA
        //
        void
        StrategyList::add_PoaThreadPool( std::size_t nbThreads)
        throw( BadParameterException )
        {

            // Check incompatible Strategy
            // ---------------------------------------------------------------------

            if ( m_has_poaThreadPerRequest )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreaded already set");
            }

            if ( m_has_poaThreadPerConnection )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreaded already set");
            }

            // Check already set Strategy
            // ---------------------------------------------------------------------
            if ( m_has_poaThreadPool )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreadPool already set");

            }


            // Check Parameters
            // ---------------------------------------------------------------------
            if ( nbThreads > 0 )
            {
                m_has_poaThreadPool = true;
                m_nbPoaThreads = nbThreads;

            }
            else
            {

                CDMW_THROW2( BadParameterException, "nbThreads", "minimum value is 1");

            }
        }



        //
        // Defines the threaded strategy for the POA
        //
        void
        StrategyList::add_PoaThreadPerRequest()
        throw( BadParameterException )
        {

            // Check incompatible Strategy
            // ---------------------------------------------------------------------

            if ( m_has_poaThreadPool )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreadPool already set");

            }

            if ( m_has_poaThreadPerConnection )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreaded already set");
            }


            // Check already set Strategy
            // ---------------------------------------------------------------------
            if ( m_has_poaThreadPerRequest )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreaded already set");
            }

            // Check Parameters
            // ---------------------------------------------------------------------
            m_has_poaThreadPerRequest = true;

        }


        //
        // Defines the threaded strategy for the POA
        //
        void
        StrategyList::add_PoaThreadPerConnection()
        throw( BadParameterException )
        {

            // Check incompatible Strategy
            // ---------------------------------------------------------------------

            if ( m_has_poaThreadPool )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreadPool already set");

            }

            if ( m_has_poaThreadPerRequest )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreaded already set");
            }


            // Check already set Strategy
            // ---------------------------------------------------------------------
            if ( m_has_poaThreadPerConnection )
            {

                CDMW_THROW2( BadParameterException, "", "PoaThreaded already set");
            }

            // Check Parameters
            // ---------------------------------------------------------------------
            m_has_poaThreadPerConnection = true;

        }



        //
        // Defines that the user want multicast
        //
        void
        StrategyList::add_multicast()
        throw( BadParameterException )
        {
          m_has_multicast = true;
        }

        //
        // Defines the port to be used by a POA
        //
        void
        StrategyList::add_PoaPortAssigned( std::size_t poaPort )
        throw( BadParameterException )
        {

            // Check incompatible Strategy
            // ---------------------------------------------------------------------
            // None


            // Check already set Strategy
            // ---------------------------------------------------------------------

            if ( m_has_poaPortAssigned )
            {

                CDMW_THROW2( BadParameterException, "", "PoaPortAssigned already set");
            }

            // Check Parameters
            // ---------------------------------------------------------------------
            if ( ! (poaPort > 0 && poaPort < 65535) )
            {

                CDMW_THROW2( BadParameterException, "", "The Port is not in the allowed range");

            }

            m_has_poaPortAssigned = true;

            m_poaPort = poaPort;

        }


        //
        // Returns the threaded strategy for the ORB
        //
        bool
        StrategyList::is_OrbThreaded() const
        throw()
        {
            return m_has_orbThreaded;

        }


        //
        // Returns the ThreadPool strategy for the POA
        //
        bool
        StrategyList::is_PoaThreadPool( std::size_t& nbThreads) const
        throw()
        {
            if ( m_has_poaThreadPool )
            {
                nbThreads = m_nbPoaThreads;
            }

            return m_has_poaThreadPool;

        }


        //
        // Returns the ThreadPool strategy for the POA
        //
        bool
        StrategyList::is_PoaThreadPerRequest() const
        throw()
        {

            return m_has_poaThreadPerRequest;

        }


        //
        // Returns the ThreadPool strategy for the POA
        //
        bool
        StrategyList::is_PoaThreadPerConnection() const
        throw()
        {

            return m_has_poaThreadPerConnection;

        }

        //
        // Returns if user want multicast
        //
        bool
        StrategyList::is_multicast() const
        throw()
        {

            return m_has_multicast;

        }


        //
        // Returns the port to be used by the POA
        //
        bool
        StrategyList::is_PoaPortAssigned( std::size_t& poaPort ) const
        throw()
        {

            if ( m_has_poaPortAssigned )
            {
                poaPort = m_poaPort;
            }

            return m_has_poaPortAssigned;

        }

    } // End namespace OrbSupport
} // End namespace Cdmw

